src/rdkafka_request.c

```diff
@@ -2355,17 +2338,118 @@ rd_kafka_resp_err_t rd_kafka_MetadataRequest(rd_kafka_broker_t *rkb,
+            RD_KAFKA_REPLYQ(rkb->rkb_rk->rk_ops, 0),
+            /* The default response handler is rd_kafka_handle_Metadata, but we
+               allow alternate handlers to be configured. */
+            resp_cb ? resp_cb : rd_kafka_handle_Metadata,
+            /* rd_kafka_handle_Metadata (the default handler) makes use of the
+               rko (if set) to forward the response to the rko's replyq. So, in
+               case the opaque value hasn't been overrriden, we set it to the
+               rko. */
+            rko ? rko : opaque);
 
         return RD_KAFKA_RESP_ERR_NO_ERROR;
 }
 
 
+/**
+ * @brief Construct MetadataRequest (does not send)
+ *
+ * \p topics is a list of topic names (char *) to request.
+ *
+ * !topics          - only request brokers (if supported by broker, else
+ *                    all topics)
+ *  topics.cnt==0   - all topics in cluster are requested
+ *  topics.cnt >0   - only specified topics are requested
+ *
+ * @param reason    - metadata request reason
+ * @param allow_auto_create_topics - allow broker-side auto topic creation.
+ *                                   This is best-effort, depending on broker
+ *                                   config and version.
+ * @param cgrp_update - Update cgrp in parse_Metadata (see comment there).
+ * @param force_racks - Force partition to rack mapping computation in
+ *                      parse_Metadata (see comment there).
+ * @param rko       - (optional) rko with replyq for handling response.
+ *                    Specifying an rko forces a metadata request even if
+ *                    there is already a matching one in-transit.
+ * @param force - 1: force a full request (including all topics and brokers)
+ *                   even if there is such a request already in flight.
+ *              - 0: check if there are multiple outstanding full requests, and
+ *                   don't send one if there is already one present.
+ *               (See note below.)
+ *
+ * If full metadata for all topics is requested (or
+ * all brokers, which results in all-topics on older brokers) and there is
+ * already a full request in transit then this function will return
+ * RD_KAFKA_RESP_ERR__PREV_IN_PROGRESS otherwise RD_KAFKA_RESP_ERR_NO_ERROR.
+ * If \p rko is non-NULL or if \p force is true, the request is sent regardless.
+ */
+rd_kafka_resp_err_t rd_kafka_MetadataRequest(rd_kafka_broker_t *rkb,
+                                             const rd_list_t *topics,
+                                             const char *reason,
+                                             rd_bool_t allow_auto_create_topics,
+                                             rd_bool_t cgrp_update,
+                                             rd_bool_t force_racks,
+                                             rd_kafka_op_t *rko,
+                                             rd_bool_t force,
+                                             void *opaque) {
```

Comment:
shall we really expose these two in this part? I think we should stick to the old one and add it when it is really required.